### DISCUSSION TITLE
How to set JsonSerializer settings for Visual Studio Language Service?

### DISCUSSION THREAD

**Discussion by CalvinAllen** [microsoft/vs-streamjsonrpc#812] [Q&A]
Long story short -

Visual Studio extension with a custom language server running out of proc (node). We've been (trying) to add support for VS 2022, and as such, upgraded a lot of the dependencies including JsonRpc/etc.

For 2017/2019, we had to set a custom contract resolver / null handling by accessing the `JsonSerializer` off of the RPC instance we're handed - something like -

```csharp
rpc.JsonSerializer.ContractResolver = new .....
rpc.JsonSerializer.NullValueHandling = Ignore;
```

With the upgrades, that `JsonSerializer` instance is no longer publicly accessible. I understand (I think) that we need to use a `JsonMessageFormatter` and set it up through that - however, I can't figure out how to utilize it since the instance of `JsonRpc` is handed to us through the `ILanguageClient` interfaces and services.

What am I missing?  Thanks!

**Comment by AArnott** [comment]
What is the type of this `rpc` variable that you're accessing `JsonSerializer` on? If that type dropped the `JsonSerializer` property, that would be a breaking change, which we don't tend to do in StreamJsonRpc except on major versions. Did you upgrade from StreamJsonRpc 1.x to 2.x?

**Reply by CalvinAllen** [reply]
Yeah we went from something like 1.5 to 2.11 (for 2022 support). Type is just 'JsonRpc'.

**Reply by AArnott** [reply]
You are correct that the `JsonMessageFormatter` is how you now get to the `JsonSerializer`. And you can't get that from the `JsonRpc` class. If the LSP APIs in VS don't expose this to you, that's something they will have to fill in since they create the `JsonRpc` object for you. Can you use Report a Problem to request this? If you link to it from here, I'll see that it gets forwarded to the right team.

I wonder... given several other LSP servers haven't hit this (AFAIK), why your server in particular requires access to customize how messages are serialized. It seems like the LSP protocol should define it, and VS should hand you one that already knows how to serialize the necessary types. What is it that you need to customize?

**Reply by CalvinAllen** [reply]
We had to apply the `CamelCasePropertyNamesContractResolver` (well, actually, a custom type based off of that, but the base type was the primary reason). I'm going to look and see if there might be anything else I can do to accomplish this, but I will create a request as well.

**Reply by CalvinAllen** [reply]
Here is the feedback request -
https://developercommunity.visualstudio.com/t/Expose-a-way-to-set-a-custom-JsonMessage/10082556

**Reply by AArnott** [reply]
Great. I've routed that ticket to the right team and will help ensure they understand the context.

**Reply by CalvinAllen** [reply]
@AArnott Thank you!
